microprofile/WFLY-13024_Integrate_MP_Fault_Tolerance_2_1.adoc

```diff
@@ -0,0 +1,87 @@
+= WFLY-12590 Integrate MicroProfile Fault Tolerance 2.1
+:author:            Radoslav Husar
+:email:             rhusar@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:keywords:          microprofile,fault-tolerance,openshift
+
+== Overview
+
+The support for Eclipse MicroProfile (MP) Fault Tolerance (FT) spec version 2.0 was already integrated for all applications
+deployed in WildFly with https://issues.jboss.org/browse/WFLY-12590[WFLY-12590].
+This describes the continuation of this effort updating the spec to version 2.1 following the Eclipse MicroProfile 3.3 release.
+
+This task incorporates upgrading SmallRye Fault Tolerance implementation to 4.1.0 (or newer) and the spec to 2.1 jars and TCK,
+removing legacy libraries and dependencies such as hystrix, commons-config, hdrhistogram, etc.
```

Comment:
After discussion with @mnovak1, I conclude that removing hystrix is much bigger change than just removing some legacy dependency. This analysis needs to contain more information about this change and how hystrix is replaced. For example how hystrix specific configuration properties are replaced in new implementation?

Reply:
> After discussion with @mnovak1, I conclude that removing hystrix is much bigger change than just removing some legacy dependency. 

The SR FT 4.x is a complete rewrite of the FT implementation , so yes, its a bigger change.

> This analysis needs to contain more information about this change and how hystrix is replaced.

That's an implementation detail, hystrix usage was removed completely from the project. If you need more details than that you would need to talk to the SR team or look at the code.

> For example how hystrix specific configuration properties are replaced in new implementation?

See the documentation in the PR, but simply put, there is no replacement since those properties are not relevant any more. The only configurable option are the thread pool sizes.

Reply:
> The SR FT 4.x is a complete rewrite of the FT implementation , so yes, its a bigger change.

@rhusar It's risky as it's new implementation. Hystrix was battle-hardened for years. Does the new implementation have the same test coverage as hystrix? If not what's is missing. 

> That's an implementation detail, hystrix usage was removed completely from the project. If you need more details than that you would need to talk to the SR team or look at the code.

It's not as it made limitations for multiple deployment and tests written for previous MP FT 2.0 RFE were written with this in mind and started to fail.

Reply:
> @rhusar It's risky as it's new implementation. 

Its not like we can choose which implementation we want to use, this is the only implementation we have that implements FT 2.1, so the question is moot. 

> Hystrix was battle-hardened for years.

On the contrary, its abandonware, outside our control and the integration has been cumbersome.

> Does the new implementation have the same test coverage as hystrix?

I believe same or better than the previous implementation. 

> It's not as it made limitations for multiple deployment and tests written for previous MP FT 2.0 RFE were written with this in mind and started to fail.

That's expected - those tests should be simply removed now.

Reply:
> Its not like we can choose which implementation we want to use, this is the only implementation we have that implements FT 2.1, so the question is moot.

I do not suggest to return back. However this RFE should be about simple update incorporating diffs between MP FT 2.0 and 2.1. Replacing Hystrix which is the core of MP FT implementation is way above of this scope.
 
> > Does the new implementation have the same test coverage as hystrix?
> 
> I believe same or better than the previous implementation.

Could you confirm it, please? (note tier 0/1 tests are handled by DEV)
If it's really ok then only new tests for integration should be added. 

> > It's not as it made limitations for multiple deployment and tests written for previous MP FT 2.0 RFE were written with this in mind and started to fail.
> 
> That's expected - those tests should be simply removed now.

Those tests should be removed or changed based on new behaviour for multiple deployment. New tests for covering it should be added. However analyses does not cover this topic well. Could you elaborate on it, please?

Reply:
Hopefully the questions were confirmed on Monday presentation. Just to summarize, removing hystrix is not above the scope, coverage is good or better as confirmed by SR team and test coverage tests, there is no special behavior when it comes to multiple deployments so nothing to cover in the proposal.